stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+      unsigned payloadTag = 0;
+      if (tag > PayloadCount
+          || !reader.readInteger(address, PayloadSize, &payloadTag)
+          || PayloadCount + payloadTag >= getNumCases()) {
+        return false;
+      }
+      *CaseIndex = PayloadCount + payloadTag;
+    } else {
+      unsigned payloadTagCount = (1U << (TagSize * 8U)) - 1;
+      unsigned maxValidTag = (NumCases - PayloadCount) / payloadTagCount + PayloadCount;
+      unsigned payloadTag = 0;
+      if (tag > maxValidTag
+          || !reader.readInteger(address, PayloadSize, &payloadTag)) {
+        return false;
+      }
+      unsigned ComputedCase = PayloadCount
+        + (tag - PayloadCount) * payloadTagCount + payloadTag;
+      if (ComputedCase >= NumCases) {
+        return false;
+      }
+      *CaseIndex = ComputedCase;
+    }
+    return true;
+  }
+};
+
+// A variable-length bitmap used to track "spare bits" for general multi-payload
+// enums.
+class BitMask {
+  unsigned size;
+  uint8_t *mask;
+public:
+  BitMask(int sizeInBytes): size(sizeInBytes) {
+    mask = (uint8_t *)malloc(size);
+    memset(mask, 0xff, size);
+  }
+  ~BitMask() {
+    free(mask);
+  }
+  // Move constructor moves ownership and zeros the src
+  BitMask(BitMask&& src) noexcept: size(src.size), mask(src.mask) {
+    src.size = 0;
+    src.mask = nullptr;
+  }
+  // Copy constructor makes a copy of the mask storage
+  BitMask(const BitMask& src) noexcept: size(src.size), mask(nullptr) {
+    mask = (uint8_t *)malloc(size);
+    memcpy(mask, src.mask, size);
+  }
+
+  void makeZero() { memset(mask, 0, size); }
+
+  bool isNonZero() const { return !isZero(); }
+
+  bool isZero() const {
+    for (unsigned i = 0; i < size; ++i) {
+      if (mask[i] != 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void complement() {
+    for (unsigned i = 0; i < size; ++i) {
+      mask[i] = ~mask[i];
+    }
+  }
+
+  int countSetBits() const {
+    static const int counter[] =
+      {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4};
+    int bits = 0;
+    for (unsigned i = 0; i < size; ++i) {
+      bits += counter[mask[i] >> 4] + counter[mask[i] & 15];
+    }
+    return bits;
+  }
+
+  int countZeroBits() const {
+    static const int counter[] =
+      {4, 3, 3, 2, 3, 2, 2, 1, 3, 2, 2, 1, 2, 1, 1, 0};
+    int bits = 0;
+    for (unsigned i = 0; i < size; ++i) {
+      bits += counter[mask[i] >> 4] + counter[mask[i] & 15];
+    }
+    return bits;
+  }
+
+  template<typename IntegerType>
+  void andMask(IntegerType value, unsigned byteOffset) {
+    andMask((void *)&value, sizeof(value), byteOffset);
+  }
+
+  void andMask(BitMask mask, unsigned offset) {
+    andMask(mask.mask, mask.size, offset);
+  }
+
+  void andNotMask(BitMask mask, unsigned offset) {
+    andNotMask(mask.mask, mask.size, offset);
+  }
+
+  // Zero all bits except for the `n` most significant ones.
+  // XXX TODO: Big-endian support?
+  void keepOnlyMostSignificantBits(int n) {
```

Comment:
Needed for computing the payload tag from the spare bits mask.  Note the comment about big-endian support.  I don't this is critical, since the corresponding compiler code lacks big-endian support as well.